File: src/main/java/me/soknight/minigram/chats/repository/ChatRepository.java
package me.soknight.minigram.chats.repository;

import me.soknight.minigram.chats.model.entity.ChatEntity;
import org.jspecify.annotations.Nullable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;

import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

public interface ChatRepository extends JpaRepository<ChatEntity, Long> {

    @EntityGraph(attributePaths = "members")
    @Query(value = """
            select distinct chat
            from ChatEntity chat
            join chat.members membership
            where membership.id.userId = :userId
            """,
            countQuery = """
            select count(distinct chat.id)
            from ChatEntity chat
            join chat.members membership
            where membership.id.userId = :userId
            """)
    Page<ChatEntity> findAllByMemberUserId(UUID userId, Pageable pageable);

    @EntityGraph(attributePaths = "members")
    @Query("""
            select distinct chat
            from ChatEntity chat
            join chat.members membership
            where chat.id = :chatId and membership.id.userId = :userId
            """)
    Optional<ChatEntity> findAccessibleById(long chatId, UUID userId);

    @Modifying
    @Query("""
            update ChatEntity chat
            set chat.lastMessageId = :messageId, chat.updatedAt = :now
            where chat.id = :chatId
            """)
    void updateLastMessageId(long chatId, @Nullable Long messageId, Instant now);

    @Modifying
    @Query("""
            update ChatEntity chat
            set chat.updatedAt = :now
            where chat.id = :chatId
            """)
    void touch(long chatId, Instant now);

    @Modifying
    @Query("""
            update ChatEntity chat
            set chat.messageSequence = chat.messageSequence + 1
            where chat.id = :chatId
            """)
    void incrementMessageSequence(long chatId);

    @Query("""
            select chat.messageSequence
            from ChatEntity chat
            where chat.id = :chatId
            """)
    long getMessageSequence(long chatId);

}

File: src/main/kotlin/me/soknight/minigram/chats/repository/ChatRepository.kt
@file:OptIn(ExperimentalUuidApi::class)

package me.soknight.minigram.chats.repository

import me.soknight.minigram.chats.domain.ChatRow
import me.soknight.minigram.chats.model.*
import me.soknight.minigram.chats.plugin.dbQuery
import org.jetbrains.exposed.v1.core.SortOrder
import org.jetbrains.exposed.v1.core.and
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.jdbc.selectAll
import org.jetbrains.exposed.v1.jdbc.update
import kotlin.time.Clock
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

class ChatRepository {

    suspend fun findAllByMemberId(userId: Uuid, page: Int, size: Int): List<ChatRow> = dbQuery {
        ChatsTable.innerJoin(ChatMembersTable)
            .selectAll()
            .where { ChatMembersTable.userId eq userId }
            .orderBy(ChatsTable.updatedAt, SortOrder.DESC)
            .limit(size).offset((page * size).toLong())
            .map { ChatEntity.wrapRow(it).toDomain() }
    }

    suspend fun findById(chatId: Long): ChatRow? = dbQuery {
        ChatEntity.findById(chatId)?.toDomain()
    }

    suspend fun findAccessibleById(chatId: Long, userId: Uuid): ChatRow? = dbQuery {
        ChatsTable.innerJoin(ChatMembersTable)
            .selectAll()
            .where { (ChatsTable.id eq chatId) and (ChatMembersTable.userId eq userId) }
            .singleOrNull()
            ?.let { ChatEntity.wrapRow(it).toDomain() }
    }

    suspend fun insert(type: ChatType, title: String?, ownerId: Uuid): ChatRow = dbQuery {
        val now = Clock.System.now()
        ChatEntity.new {
            this.type          = type.name
            this.title         = title
            this.ownerId       = ownerId
            this.messageSeq    = 0
            this.lastMessageId = null
            this.createdAt     = now
            this.updatedAt     = now
        }.toDomain()
    }

    suspend fun updateTitle(chatId: Long, newTitle: String?): Unit = dbQuery {
        ChatsTable.update({ ChatsTable.id eq chatId }) {
            it[title]     = newTitle
            it[updatedAt] = Clock.System.now()
        }
    }

    suspend fun delete(chatId: Long): Unit = dbQuery {
        ChatEntity.findById(chatId)?.delete()
    }

    suspend fun incrementMessageSequence(chatId: Long): Long = dbQuery {
        val entity = ChatEntity.findById(chatId) ?: error("Chat $chatId not found")
        entity.messageSeq += 1
        entity.messageSeq
    }

    suspend fun updateLastMessageId(chatId: Long, messageId: Long?): Unit = dbQuery {
        ChatsTable.update({ ChatsTable.id eq chatId }) {
            it[lastMessageId] = messageId
            it[updatedAt]     = Clock.System.now()
        }
    }
}
